src/math/MathUtils.js

```diff
@@ -291,7 +291,32 @@ function normalize( value, array ) {
 
 }
 
-
+const MathUtils = {
+	DEG2RAD: DEG2RAD,
+	RAD2DEG: RAD2DEG,
+	generateUUID: generateUUID,
+	clamp: clamp,
+	euclideanModulo: euclideanModulo,
+	mapLinear: mapLinear,
+	inverseLerp: inverseLerp,
+	lerp: lerp,
+	damp: damp,
+	pingpong: pingpong,
+	smoothstep: smoothstep,
+	smootherstep: smootherstep,
+	randInt: randInt,
+	randFloat: randFloat,
+	randFloatSpread: randFloatSpread,
+	seededRandom: seededRandom,
+	degToRad: degToRad,
+	radToDeg: radToDeg,
+	isPowerOfTwo: isPowerOfTwo,
+	ceilPowerOfTwo: ceilPowerOfTwo,
+	floorPowerOfTwo: floorPowerOfTwo,
+	setQuaternionFromProperEuler: setQuaternionFromProperEuler,
+	normalize: normalize,
+	denormalize: denormalize
+};
```

Comment:
Hey, exporting as plain object breaks tree-shaking, we can't have that.

Having `export * as` is fine as it gets transpiled by rollup in the build step.

Reply:
Let's look into that.  There are a number of objects of this style.  This is what it gets trans-piled to.  I have a tree-shake confirmation project published.

This is being done in 2 parts.  https://github.com/mrdoob/three.js/issues/25185

Reply:
Update: https://github.com/epreston/three-treeshake-coverage 

Please pretend you don't see the other bits.  Need an hour to cleanup my changes and commit them.

Reply:
Long story short:

> Having `export * as` is fine as it gets transpiled by rollup in the build step.

Yes, it becomes something, which is very similar to this, before it gets distributed and the final bundler (client application) does the shake.  If we want to improve the shake, we have API changes.  If we want to retain the API, distribute nearly the same bundle, and enforce 2018, we change the code.

Currently the tools are configured for 2020.  Setting the tools to 2018 and enforcing it is the first step so we actually improve the tree-shake instead of masking it.

Reply:
As a confirmation take a look at what we ship now using the 2020 syntax.  I'm going to use the middle size one to reduce the reading.

The old 2020 line:

```js
export * as AnimationUtils from './animation/AnimationUtils.js';
```

Is currently being translated to:

```js
var AnimationUtils = /*#__PURE__*/Object.freeze({
	__proto__: null,
	arraySlice: arraySlice,
	convertArray: convertArray,
	flattenJSON: flattenJSON,
	getKeyframeOrder: getKeyframeOrder,
	isTypedArray: isTypedArray,
	makeClipAdditive: makeClipAdditive,
	sortedArray: sortedArray,
	subclip: subclip
});
```

This has been updated to:

```js
const AnimationUtils = {
	arraySlice: arraySlice,
	convertArray: convertArray,
	isTypedArray: isTypedArray,
	getKeyframeOrder: getKeyframeOrder,
	sortedArray: sortedArray,
	flattenJSON: flattenJSON,
	subclip: subclip,
	makeClipAdditive: makeClipAdditive
};
```

It gets translated to:

```js
const AnimationUtils = {
	arraySlice: arraySlice,
	convertArray: convertArray,
	isTypedArray: isTypedArray,
	getKeyframeOrder: getKeyframeOrder,
	sortedArray: sortedArray,
	flattenJSON: flattenJSON,
	subclip: subclip,
	makeClipAdditive: makeClipAdditive
};
```

Keeping it clean, the final bundler works its magics.  This is what I hope to demonstrate in the confirmation project.

Reply:
I'm writing this all out because you are right, except if we use this particular 2020 shorthand syntax, which doesn't help the cause.

Reply:
> If we want to improve the shake, we have API changes.

Actually you're correct, the object in the final bundle doesn't get tree-shaken because it's an object, and budnlers can't tree-shake that. The only way to make it tree-shakable would be to do `export * from './animation/AnimationUtils.js';`, but I'm not sure maintainers want to go with this with this breaking change.

> except if we use this particular 2020 shorthand syntax, which doesn't help the cause.
There is no problem with this syntax, as it doesn't end up in production. What #25185 is all about is Optional chaining and Nullish coalescing (and of course class fields from ES2022).

I guess we can enforce ES2020 and disable just Optional chaining and Nullish coalescing.

Reply:
Unrelated but...

Could be nice if the CI could check if a PR makes the build bigger/smaller and also if the tree-shaked lib is bigger/smaller.

Reply:
@mrdoob its on the way.  Also circular reference check.

Reply:
> I guess we can enforce ES2020 and disable just Optional chaining and Nullish coalescing.

At the moment, that is not possible.  With the agreed 2018 direction, the zero impact to API and treeshake, this is the best we have.

Reply:
Thanks @epreston! That would be cool. You can get the tree-shaken lib size from `test/treeshake/index.bundle.min.js`.

Reply:
Also maybe it can report test coverage? I think many large projects use bots that report both build size and test coverage.

Reply:
I really believe in 50% best case reduction if we make some really careful and clever moves.

Reply:
> At the moment, that is not possible. With the agreed 2018 direction, the zero impact to API and treeshake, this is the best we have.

Why is it not possible?

Reply:
> disable just Optional chaining

I don't think anyone wants to disable "Optional Chaining" that would drop a massive amount of code from the library and make things very robust.  Seems like half of most classes are just checking for null / undefined.



> Nullish coalescing

Yeah, that's hard to read, causes bugs, and I can't spell it correctly without looking it up.

Reply:
Quick heads up, "circular reference checks" are not going to win me any friends.  Nodes has 22.  Analysis posted here: #25240

Reply:
> I don't think anyone wants to disable "Optional Chaining" that would drop a massive amount of code from the library and make things very robust. Seems like half of most classes are just checking for null / undefined.

We have already disabled it and all of its usages.

> Nodes has 22. Analysis posted here: https://github.com/mrdoob/three.js/issues/25240

The solution is also posted there -- I will file a PR soon.

Reply:
> I don't think anyone wants to disable "Optional Chaining" that would drop a massive amount of code from the library and make things very robust. Seems like half of most classes are just checking for null / undefined.

By [Optional chaning](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) I mean `?.` and currently it's only used on `GLTFExporter.js`.
Nullish coalescing is not used anywhere in three.js.

Reply:
> > At the moment, that is not possible. With the agreed 2018 direction, the zero impact to API and treeshake, this is the best we have.
> 
> Why is it not possible?

Not exposed as a rule, looks like it requires a custom eslint parser to implement.

Reply:
Ah yeah, well we could use the rules from this plugin: https://github.com/keithamus/eslint-plugin-escompat

Reply:
I see this as the best solution, because as mentioned here https://github.com/mrdoob/three.js/issues/25185#issuecomment-1422059604, the import/exports are not included in the enforcing.

Reply:
That looks promising for the next step.  Thank you. 

The plan is to relax back to 2020 parsing with a set of rules to strip it back that we can turn on and off.

Can you repost that on https://github.com/mrdoob/three.js/issues/25185 so we don't lose it ?

Reply:
> I see this as the best solution, because as mentioned here [#25185 (comment)](https://github.com/mrdoob/three.js/issues/25185#issuecomment-1422059604), the import/exports are not included in the enforcing.

Which solution ?  There's a few more bits to consider to enforce 2018, without setting the tools to enforce, but add extra config to "sort of" enforce 2018.

Reply:
ES2020 + `no-nullish-coalescing` and `no-optional-chaining` would be the best solution.

Reply:
The project you linked is for things that can't be polyfilled and trans-piled.  Keep in mind, Babel isn't in the toolchain.  So we don't have full polyfill and trans-pile.  Your idea has changed my mind about how much effort the next step will take.   I think it is awesome and we actually completely agree if you read what I wrote yesterday.

People keep confusing what is in the bundle vs what code is maintained.  It really is productive to have a separate discussion on the code we read and share.  Then, have another conversation about what is included in the bundle how that is presented.

At the moment, we don't bundle to a language standard but a browser support standard.  The language features used look like a rough mixture of es6 which changes every quarter.  Talking about the bundle in terms of language features makes little sense when its defined in terms of browser support.

The issue I see coming up will be about perceptions, assumptions, and confusion.  We have a standard, but we deviate from the standard, but something will fix it, then finally realising that is not what shipped.

I want to make people happy so they can produce the exact same output and remove 20 lines from the project.  If that needs to be a birds nest of tooling config, so be it.  We can make it so people can use a few 2020 features in a context of 2018.  Not keen on doing the work then having to remediate what has crept in.  Its easier to test and rationalise about with a stable base going forward.

Reply:
> ES2020 + `no-nullish-coalescing` and `no-optional-chaining` would be the best solution.

That is not a complete solution or a very wise one.

Reply:
I'm sorry.  That was too short of a response.  

My goal was to patiently explain why the assumption on tree-shaking was wrong.  When this came up I felt overwhelmed about going into what's missing.  I had the same ideas and learned the hard way.

I'll make it up to you.  This is not who I want to be.